addon/utils/fetch-network-interface.js

```diff
@@ -0,0 +1,30 @@
+import { HTTPFetchNetworkInterface, printAST } from 'apollo-client';
+import fetch from 'fetch';
+import Ember from 'ember'
+
+const { RSVP: { Promise } } = Ember;
+
+export default class FetchNetworkInterface extends HTTPFetchNetworkInterface {
+  fetchFromRemoteEndpoint({ request, options }) {
+    var variables = request.variables;
+    var query = printAST(request.query);
+
+    return new Promise( (resolve, reject) => {
+      return fetch(this._uri, {
+        ...this._opts,
+        method: 'POST',
+        body: JSON.stringify({variables, query}),
+        ...options,
+        headers: {
+          Accept: '*/*',
+          'Content-Type': 'application/json',
+          ...options.headers
+        }
+      }).then((response) =>{
+        resolve(response);
+      }).catch((response) =>{
+        reject(response);
+      })
+    });
+  }
+} 
```

Comment:
In Apollo Client 2.0, [network interfaces are deprecated in favor of Apollo Link](https://github.com/apollographql/apollo-client/blob/master/Upgrade.md#details). I'm a bit concerned that adding this custom network interface now makes the upgrade harder as this may have to be largely rewritten. Is that accurate?

Additionally, there are no tests at all for this new interface, and I don't think I should merge it without some decent test coverage.

Reply:
@bgentry but the master branch still not have Apollo client. I think that this request is needless for this PR.

Reply:
If we are planning to migrate to Apollo Client 2.0 right away, I agree with Blake about introducing a not tested "stillborn" code. This can impact the migration of this addon and maybe the migration of the applications that rely on it.

On the other hand, a Fastboot support is a must have, so I think this is a matter of how soon do we plan to migrate to 2.0.

@bgentry What are your plans about it? Should we plan something on #54 to see the amount of work we are going to face?

Reply:
I still haven't spent enough time looking into the 2.0 migration to know how much work it will be. That's the part I'm unsure about.

However, I'm very sure that I can't merge this without sufficient test coverage of the NetworkInterface. It's core code that every user depends on, and by switching from the default Apollo implementation to our own, we're losing test coverage of it.

Reply:
I think the most important thing is the community's greatest need right now, which is about fastboot support. But it would be good to start migrating to version 2.0, but without leaving the community waiting

Reply:
@bgentry Testing this `util`, is this PR will be ready for merge?

Reply:
I think we are missing the point here. The question is: _why do we need a new network interface?_ Apollo Client already ships with `HTTPFetchNetworkInterface `, [that rely on Fetch API](https://github.com/apollographql/apollo-client/blob/v1.9.0/src/transport/networkInterface.ts#L186-L196). It's the [network interface returned by `createNetworkInterface` function](https://github.com/apollographql/apollo-client/blob/v1.9.0/src/transport/networkInterface.ts#L290). 

Changing `ember-network` to `ember-fetch` should be enough to be in compliance with Fastboot.

@marucjmar Am I missing something?

Reply:
I agree

Reply:
@viniciussbs  HTTPFetchNetworkInterface seems to be based on [whatwg-fetch](https://github.com/apollographql/apollo-client/blob/v1.9.0/src/transport/networkInterface.ts#L1) which is a [polyfill](https://github.com/github/fetch)

Reply:
I thought about using apollo-fetch instead of creating a new network interface, with something like [this](https://github.com/balaf/ember-apollo-client/commit/c85de9b5539eea1ac4953c57478ff36b2ef04ce8) but I can't get this to work yet

Reply:
@balaf Looks like [`whatwg-fetch` does nothing if there's a global `fetch`](https://github.com/github/fetch/blob/master/fetch.js#L4-L6), so `ember-fetch` might do the trick if loaded beforehand. It's a matter of testing the new version of `ember-apollo-client` now that #59 is merged.

Who volunteers to write a Fastboot compatible app to fetch data from a [public GraphQL API](https://github.com/APIs-guru/graphql-apis#demonstration-only-apis)?

@bgentry #59 was released or it's only on master?

Reply:
I get XMLHttpRequest not defined error in my app when I turn fastboot on

Reply:
What branch. @balaf?

Reply:
@balaf Are you facing it after #59 ? Have you checked if this updated was in the last release? Maybe you need to test it on master.

Reply:
Yes, #59 is in and I am facing XMLHttpRequest errors when I run fastboot. My app in not in a public github repo just yet (but it will be in the next couple of weeks).